File: samples/akka-sample-cluster-scala/src/main/scala/sample/cluster/stats/StatsClient.scala
package sample.cluster.stats

import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Behaviors

import scala.concurrent.duration._

object StatsClient {

  sealed trait Event
  private case object Tick extends Event
  private case class ServiceResponse(result: StatsService.Response) extends Event

  def apply(service: ActorRef[StatsService.ProcessText]): Behavior[Event] =
    Behaviors.setup { ctx =>
      Behaviors.withTimers { timers =>
        timers.startTimerWithFixedDelay(Tick, Tick, 2.seconds)
        val responseAdapter = ctx.messageAdapter(ServiceResponse.apply)

        Behaviors.receiveMessage {
          case Tick =>
            ctx.log.info("Sending process request")
            service ! StatsService.ProcessText("this is the text that will be analyzed", responseAdapter)
            Behaviors.same
          case ServiceResponse(result) =>
            ctx.log.info("Service result: {}", result)
            Behaviors.same
        }
      }
    }

}

File: samples/akka-sample-cluster-java/src/main/java/sample/cluster/stats/StatsClient.java
package sample.cluster.stats;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.Behaviors;

import java.time.Duration;


public final class StatsClient {

  interface Event {}
  private enum Tick implements Event {
    INSTANCE
  }
  private static class ServiceResponse implements Event {
    public final StatsService.Response result;
    public ServiceResponse(StatsService.Response result) {
      this.result = result;
    }
  }

  public static Behavior<Event> create(ActorRef<StatsService.ProcessText> service) {
    return Behaviors.setup(context ->
      Behaviors.withTimers(timers -> {
        timers.startTimerWithFixedDelay(Tick.INSTANCE, Tick.INSTANCE, Duration.ofSeconds(2));
        ActorRef<StatsService.Response> responseAdapter =
            context.messageAdapter(StatsService.Response.class, ServiceResponse::new);

        return Behaviors.receive(Event.class)
            .onMessageEquals(Tick.INSTANCE, () -> {
              context.getLog().info("Sending process request");
              service.tell(new StatsService.ProcessText("this is the text that will be analyzed", responseAdapter));
              return Behaviors.same();
            }).onMessage(ServiceResponse.class, response -> {
              context.getLog().info("Service result: {}", response.result);
              return Behaviors.same();
            }).build();
      })
    );
  }

}
